File: spawner.py
import os
import shutil
import subprocess
import secrets
from pathlib import Path
import config
import db

def get_user_dir(username: str) -> Path:
    return Path(config.BASE_DIR) / username

def setup_user_env(username: str) -> bool:
    user_dir = get_user_dir(username)
    os.makedirs(user_dir, exist_ok=True)
    
    venv_dir = user_dir / ".venv"
    if not venv_dir.exists():
        # 1. Create venv using uv
        cmd_venv = ["uv", "venv", "--python", config.PYTHON_VERSION, str(venv_dir)]
        res_venv = subprocess.run(cmd_venv, capture_output=True, text=True)
        if res_venv.returncode != 0:
            print(f"Error creating venv for {username}: {res_venv.stderr}")
            return False
        
        # 2. Install jupyterlab using uv pip inside the venv
        uv_bin = venv_dir / "bin" / "uv"
        # If uv is not copied to venv, use active system uv with --venv option or direct path
        # Simplest: use uv pip install with --python pointing to venv interpreter
        python_bin = venv_dir / "bin" / "python"
        cmd_install = ["uv", "pip", "install", f"jupyterlab=={config.JUPYTERLAB_VERSION}", "--python", str(python_bin)]
        res_install = subprocess.run(cmd_install, capture_output=True, text=True)
        if res_install.returncode != 0:
            print(f"Error installing jupyterlab for {username}: {res_install.stderr}")
            return False
            
    return True

def get_session_name(username: str) -> str:
    return f"hub_{username}"

def is_session_running(username: str) -> bool:
    session_name = get_session_name(username)
    cmd = ["tmux", "has-session", "-t", session_name]
    res = subprocess.run(cmd, capture_output=True)
    return res.returncode == 0

def spawn_session(username: str, port: int, token: str) -> bool:
    if not setup_user_env(username):
        return False
        
    if is_session_running(username):
        stop_session(username)
        
    session_name = get_session_name(username)
    user_dir = get_user_dir(username)
    jupyter_bin = user_dir / ".venv" / "bin" / "jupyter"
    
    # Construct the jupyter lab startup command
    # cd into user_dir first so JupyterLab terminal opens in user's directory
    jupyter_cmd = (
        f"cd {user_dir} && "
        f"exec {jupyter_bin} lab "
        f"--ip=127.0.0.1 "
        f"--port={port} "
        f"--IdentityProvider.token={token} "
        f"--no-browser "
        f"--notebook-dir={user_dir}"
    )
    
    # Spawn via tmux
    # -d means detached
    # -s session_name
    # command is executed inside the shell in tmux session
    cmd = ["tmux", "new-session", "-d", "-s", session_name, jupyter_cmd]
    res = subprocess.run(cmd, capture_output=True, text=True)
    if res.returncode != 0:
        print(f"Error spawning tmux session for {username}: {res.stderr}")
        return False
        
    return True

def stop_session(username: str) -> bool:
    if not is_session_running(username):
        return True
    session_name = get_session_name(username)
    cmd = ["tmux", "kill-session", "-t", session_name]
    res = subprocess.run(cmd, capture_output=True)
    return res.returncode == 0

def cleanup_user_files(username: str):
    user_dir = get_user_dir(username)
    if user_dir.exists():
        shutil.rmtree(user_dir)

def get_next_port() -> int:
    used_ports = set(db.get_used_ports())
    for port in range(config.JUPYTER_PORT_START, config.JUPYTER_PORT_END + 1):
        if port not in used_ports:
            return port
    return None

def sync_sessions():
    """Startup reconciliation: check all DB users against running tmux sessions and update tokens if stopped."""
    users = db.list_users()
    for user in users:
        username = user['username']
        if not is_session_running(username) and user['token'] is not None:
            # Session died while hub was offline, clear token in DB
            db.update_token(username, None)

File: src/spawner.ts
import { existsSync, mkdirSync, readdirSync, rmSync } from "node:fs";
import { join, resolve } from "node:path";
import { BASE_DIR, PYTHON_VERSION, JUPYTERLAB_VERSION, JUPYTER_PORT_START, JUPYTER_PORT_END, CODE_SERVER_PORT_OFFSET } from "./config";
import { listUsers, updateToken, getUsedPorts } from "./db";

export function validateUsername(username: string): void {
  if (!/^[a-zA-Z0-9_-]+$/.test(username)) {
    throw new Error(`Invalid username: ${username}`);
  }
}

export function getUserDir(username: string): string {
  validateUsername(username);
  const resolvedBase = resolve(BASE_DIR);
  const resolvedUser = resolve(join(resolvedBase, username));
  if (!resolvedUser.startsWith(resolvedBase)) {
    throw new Error(`Path traversal attempt detected: ${username}`);
  }
  return resolvedUser;
}

export async function setupUserEnv(username: string): Promise<boolean> {
  validateUsername(username);
  const userDir = getUserDir(username);
  mkdirSync(userDir, { recursive: true });

  const venvDir = join(userDir, ".venv");

  // Check if existing venv has a working python binary (guards against stale venvs
  // from old TLJH installations where symlinks point to deleted paths).
  if (existsSync(venvDir)) {
    const pythonBin = join(venvDir, "bin", "python");
    const checkProc = Bun.spawn([pythonBin, "--version"], {
      stdout: "ignore",
      stderr: "ignore",
    });
    const checkExit = await checkProc.exited;
    if (checkExit !== 0) {
      console.warn(`Broken venv detected for ${username} (python check failed) — removing and recreating`);
      rmSync(venvDir, { recursive: true, force: true });
    }
  }

  if (!existsSync(venvDir)) {
    // Create venv
    const venvProc = Bun.spawn(["uv", "venv", "--python", PYTHON_VERSION, venvDir], {
      stdout: "ignore",
      stderr: "pipe",
    });
    const stderrText = await new Response(venvProc.stderr).text();
    const exitCode = await venvProc.exited;
    if (exitCode !== 0) {
      console.error(`Error creating venv for ${username}: ${stderrText}`);
      return false;
    }

    // Install jupyterlab
    const pythonBin = join(venvDir, "bin", "python");
    const installProc = Bun.spawn(
      ["uv", "pip", "install", `jupyterlab==${JUPYTERLAB_VERSION}`, "ipykernel", "--python", pythonBin],
      { stdout: "ignore", stderr: "pipe" }
    );
    const stderrTextInstall = await new Response(installProc.stderr).text();
    const installExit = await installProc.exited;
    if (installExit !== 0) {
      console.error(`Error installing jupyterlab for ${username}: ${stderrTextInstall}`);
      return false;
    }
  }
  return true;
}

function getSessionName(username: string): string {
  validateUsername(username);
  return `hub_${username}`;
}

export async function isSessionRunning(username: string): Promise<boolean> {
  validateUsername(username);
  const proc = Bun.spawn(["tmux", "has-session", "-t", getSessionName(username)], {
    stdout: "ignore",
    stderr: "ignore",
  });
  return (await proc.exited) === 0;
}

export async function spawnSession(username: string, port: number, token: string): Promise<boolean> {
  validateUsername(username);
  if (typeof port !== "number" || !Number.isInteger(port) || port < JUPYTER_PORT_START || port > JUPYTER_PORT_END) {
    throw new Error(`Invalid port: ${port}`);
  }
  if (!/^[a-zA-Z0-9_-]+$/.test(token)) {
    throw new Error("Invalid token format");
  }
  if (!(await setupUserEnv(username))) return false;

  if (await isSessionRunning(username)) {
    await stopSession(username);
  }

  const sessionName = getSessionName(username);
  const userDir = getUserDir(username);
  const jupyterBin = join(userDir, ".venv", "bin", "jupyter");
  const codeServerPort = port + CODE_SERVER_PORT_OFFSET;

  const jupyterCmd = `cd ${userDir} && exec ${jupyterBin} lab --ip=0.0.0.0 --port=${port} --IdentityProvider.token=${token} --no-browser --notebook-dir=${userDir}`;

  // Start tmux with jupyter in window 0
  const proc = Bun.spawn(["systemd-run", "--user", "--scope", "tmux", "new-session", "-d", "-s", sessionName, "-n", "jupyter", jupyterCmd], {
    stdout: "ignore",
    stderr: "pipe",
  });
  const stderrText = await new Response(proc.stderr).text();
  const exitCode = await proc.exited;
  if (exitCode !== 0) {
    console.error(`Error spawning tmux session for ${username}: ${stderrText}`);
    return false;
  }

  // Spawn local code-server in window 1
  try {
    const codeServerCmd = `PASSWORD=${token} code-server --bind-addr 127.0.0.1:${codeServerPort} --auth password --disable-telemetry --user-data-dir=${userDir}/.code-server`;
    const winProc = Bun.spawn(["tmux", "new-window", "-t", `${sessionName}:1`, "-n", "vscode", "sh", "-c", codeServerCmd], {
      stdout: "ignore",
      stderr: "pipe",
    });
    const winStderr = await new Response(winProc.stderr).text();
    const winExit = await winProc.exited;
    if (winExit !== 0) {
      console.warn(`Warning: Failed to spawn local code-server window for ${username}: ${winStderr}`);
    }
  } catch (e: any) {
    console.warn(`Warning: Local code-server could not be initialized: ${e.message}`);
  }

  return true;
}

export async function stopSession(username: string): Promise<boolean> {
  validateUsername(username);
  if (!(await isSessionRunning(username))) return true;
  const proc = Bun.spawn(["tmux", "kill-session", "-t", getSessionName(username)], {
    stdout: "ignore",
    stderr: "ignore",
  });
  return (await proc.exited) === 0;
}

export function cleanupUserFiles(username: string): void {
  validateUsername(username);
  const userDir = getUserDir(username);
  if (existsSync(userDir)) {
    rmSync(userDir, { recursive: true, force: true });
  }
}

export function getNextPort(): number | null {
  const used = new Set(getUsedPorts());
  for (let port = JUPYTER_PORT_START; port <= JUPYTER_PORT_END; port++) {
    if (!used.has(port)) return port;
  }
  return null;
}

export async function syncSessions(): Promise<void> {
  const users = listUsers();
  for (const user of users) {
    if (!(await isSessionRunning(user.username)) && user.token !== null) {
      updateToken(user.username, null);
    }
  }
}
